### DISCUSSION TITLE
How to deploy a plugin in development

### DISCUSSION THREAD

**Discussion by JonasDoe** [ether/etherpad-lite#5270] [Q&A]
Hey,

I'm just getting started with the whole plugin topic and I'm wondering what the most effective way is to run my plugin for testing. So two questions arose for me:
- How can I add the changes in my plugin to the etherpad? At the moment I'm always copying my whole plugin project folder into `node_modules` before running etherpad, but that seems rather inefficient. My first idea was to create a link (or symlink) from my project into the etherpad's `node_modules` folder, but neither of those are actually regarded (maybe because some kind of merging with `src/node_modules` takes place? n/c As I said, copying works).
- Is there some kind of hot-swapping to redeploy changes of my plugin without restarting the whole etherpad server and reload the website afterwards? I guess not, but there's nothing wrong with asking I guess. :)

Thanks it advance!

**Comment by webzwo0i** [comment]
You could give https://github.com/storytouch/ep_webpack a try as it does auto-updates afaik. Ensure maxAge in settings.json is 0.

Regarding node_modules: iirc use `npm link` in plugin's directory to make npm aware of your plugin (outside Etherpad's directory) and then you can simply `npm link plugin-name`

**Reply by JonasDoe** [reply]
Thanks for the quick responses so far! `ep_webpack` looks good, I'll try my luck with that one! Regarding the linking, it doesn't work out: Running the etherpad with the `ep_print` plugin linked in fails with `Cannot find module 'ep_etherpad-lite/node/eejs/'`.

**Comment by JohnMcLear** [comment]
For Server side changes a restart is required afaik, for client(html/css/js) it is not assuming minify is false and maxAge is 0

**Reply by JonasDoe** [reply]
I see, thank you!

**Comment by rhansen** [comment]
Here's what I do:

```shell
ETHERPAD_DIR=/path/to/etherpad-lite
PLUGIN_DIR=/path/to/ep_foo
(cd "$PLUGIN_DIR" && npm i)
ln -s "$ETHERPAD_DIR/src" "$PLUGIN_DIR/node_modules/ep_etherpad-lite"
ln -s "$PLUGIN_DIR" "$ETHERPAD_DIR/node_modules"
(cd "$ETHERPAD_DIR/src" && npm i)
"$ETHERPAD_DIR/src/bin/fastRun.sh"
```

**Reply by JonasDoe** [reply]
Thanks for the script! For some reason, when testing it with `ep_print`it fails for me because there's a conflict between peer resolving with `ep_etherpad-lite@undefined` at the plugin folder and peer `ep_etherpad-lite@">=1.8.6" from the root project`. My guess was to create link from the the `etherpad-lite/src` named `ep_etherpad_lite` in `ep_print/node_modules` folder, but it did not help.

**Reply by rhansen** [reply]
> peer `ep_etherpad-lite@">=1.8.6" from the root project`

Root project? Something funky is going on. Does `package.json` or `package-lock.json` exist in `$ETHERPAD_DIR`? If so, delete them.

**Reply by JonasDoe** [reply]
They don't exist, I've already learnt this lession when I first tried to install a plugin. :) That said, I only managed once to get this error again when retrying. Now most of the time the etherpad fails with 
```
DEBUG] AbsolutePaths - /Users/xyz/etherpad/ep_print/node_modules/ep_etherpad-lite does not end with "src"
[INFO] console - [2021-11-11 10:01:45.878] [ERROR] AbsolutePaths - Could not identity Etherpad base path in this darwin installation in "/Users/xyz/ep_print/node_modules/ep_etherpad-lite"
```
, which is really odd because that's not the error I got the other day and I've no idea what I've actually changed since then. On the other hand, if the run script really relies on a folder named `src`, I can understand it won't like the alias' name. But anyway, your script is rather helpful to understand where symlinks are required, so I guess I can tinker a bit with that!

**Reply by rhansen** [reply]
Run Etherpad from `$ETHERPAD_DIR`, not from `$PLUGIN_DIR/node_modules/ep_etherpad-lite`.

**Reply by JonasDoe** [reply]
Well, it's run from the last line of your script: `"$ETHERPAD_DIR/src/bin/fastRun.sh"`
